apps/api_web/lib/api_web/event_stream/canary.ex

```diff
@@ -21,7 +21,8 @@ defmodule ApiWeb.EventStream.Canary do
   end
 
   @impl true
-  def terminate(:shutdown, notify_fn), do: notify_fn.()
+  def terminate(:shutdown, notify_fn) when is_function(notify_fn), do: notify_fn.()
+  def terminate(:shutdown, _), do: @default_notify_fn.()
```

Comment:
I assume this is happening because in practice, the GenServer is started like:
```elixir
Canary.start_link([])
```

Rather than specifying the `default_notify_fn` twice, what about changing the `start_link` to (correctly) treat the arguments as a Keyword list?
```elixir
def start_link(args) do
  notify_fn = case Keyword.fetch(args, :notify_fn) do
    {:ok, fun} when is_function(fun, 0) -> fun
    {:ok, not_fun} -> raise ArgumentError, "expect function/0 for notify_fn, got #{inspect(not_fun)}"
    _ -> @default_notify_fn
  end
  
  GenServer.start_link(__MODULE__, notify_fn, [])
end
```

Reply:
Do we want to throw an exception in `start_link` or should we return `{:stop, reason}` from `init` instead? https://hexdocs.pm/elixir/1.14.0/GenServer.html#c:init/1 According to the docs:
> Returning {:stop, reason} will cause [start_link/3](https://hexdocs.pm/elixir/1.14.0/GenServer.html#start_link/3) to return {:error, reason} and the process to exit with reason reason without entering the loop or calling [terminate/2](https://hexdocs.pm/elixir/1.14.0/GenServer.html#c:terminate/2).